File: calculator.py
"""
CTD Filter Delay Calculator
============================
Calculates the preset timer delays for each filter actuation on a CTD cast.

Workflow:
  T=0: timers activated (CTD not yet in water)
  [initial margin] → downcast through Type 1 filter stops → bottom turn → upcast through Type 2 filter stops

Travel model: symmetric trapezoidal velocity profile (0 → v_max → 0) at every stop.
"""

import math
from dataclasses import dataclass, field
from typing import List


# ---------------------------------------------------------------------------
# Travel-time model
# ---------------------------------------------------------------------------

def travel_time(distance_m: float, v_max: float, accel: float) -> float:
    """
    Time to travel `distance_m` with symmetric acceleration/deceleration ramps.

    Trapezoidal profile:
      - If distance allows reaching v_max: ramp-up + cruise + ramp-down
      - Otherwise: triangular profile (never reaches v_max)

    Parameters
    ----------
    distance_m : float  Distance in metres (non-negative)
    v_max      : float  Maximum speed in m/s
    accel      : float  Acceleration / deceleration magnitude in m/s²

    Returns
    -------
    float  Travel time in seconds
    """
    if distance_m <= 0:
        return 0.0

    # Distance covered during one full ramp (0 → v_max or v_max → 0)
    ramp_dist = (v_max ** 2) / (2 * accel)
    full_ramp_dist = 2 * ramp_dist  # both ramps together

    if distance_m >= full_ramp_dist:
        # Trapezoidal: ramp up + cruise + ramp down
        ramp_time = v_max / accel          # time for one ramp
        cruise_dist = distance_m - full_ramp_dist
        cruise_time = cruise_dist / v_max
        return 2 * ramp_time + cruise_time
    else:
        # Triangular: v_peak = sqrt(a * d), t = 2 * v_peak / a
        v_peak = math.sqrt(accel * distance_m)
        return 2 * v_peak / accel


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------

@dataclass
class FilterStop:
    """A single filter actuation stop."""
    filter_id: str
    filter_type: int          # 1 = downcast, 2 = upcast
    depth_m: float
    arrival_time_s: float     # exact cumulative time from T=0
    departure_time_s: float   # exact time CTD leaves this depth
    preset_delay_s: float     # = arrival_time_s + safety_margin (what you program into the device)
    filter_duration_s: float = 900.0  # timeout for THIS filter (excludes homing/buffer)


@dataclass
class CastEvent:
    """Any event in the cast timeline (for the chart)."""
    label: str
    time_s: float
    depth_m: float
    event_type: str  # 'filter_start', 'filter_end', 'bottom', 'surface', 'travel'


@dataclass
class CastParams:
    max_depth_m: float
    type1_depths: List[float]        # downcast filter depths, shallow→deep
    type2_depths: List[float]        # upcast filter depths, deep→shallow (expected/predicted)
    v_max: float = 1.0               # m/s
    accel: float = 0.1               # m/s²
    filter_duration_s: float = 900.0 # global default filter timeout (fallback)
    safety_margin_s: float = 300.0   # global default safety margin (fallback)
    n_bottles_bottom: int = 0
    n_bottles_dcm: int = 0
    n_bottles_surface: int = 0
    bottle_close_time_s: float = 5.0
    # Per-filter overrides (lists, one entry per filter in order). Empty = use global default.
    type1_filter_durations: List[float] = field(default_factory=list)
    type2_filter_durations: List[float] = field(default_factory=list)
    type1_safety_margins: List[float] = field(default_factory=list)
    type2_safety_margins: List[float] = field(default_factory=list)
    # New timing parameters (global)
    deploy_delay_s: float = 0.0       # T=0 to water entry (CTD on deck before descent)
    bottom_dwell_s: float = 5.0       # time spent at max depth before ascending
    homing_time_s: float = 30.0       # instrument spin-up after timer fires
    actuation_buffer_s: float = 10.0 # loading time after homing before actual pumping


# ---------------------------------------------------------------------------
# Core delay computation
# ---------------------------------------------------------------------------

def compute_delays(p: CastParams):
    """
    Compute preset delays for all filters and build the full cast timeline.

    Timing model per filter:
      - CTD arrives at depth           → arrival_time_s
      - After safety_margin_s, timer fires → preset_delay_s = arrival + margin
      - After timer fires: homing_time → activation_buffer → filter_duration (all sequential dwell)
      - CTD departs                    → departure_time_s = preset + homing + buffer + filter_dur

    Returns
    -------
    filter_stops : list[FilterStop]
    timeline     : list[CastEvent]   – all events for the depth-time chart
    total_cast_time_s : float
    """
    # Validate inputs
    if not p.type1_depths and not p.type2_depths:
        raise ValueError("At least one filter depth must be provided.")

    t1 = sorted(p.type1_depths)   # shallow → deep
    t2 = sorted(p.type2_depths, reverse=True)  # deep → shallow

    filter_stops: List[FilterStop] = []
    timeline: List[CastEvent] = []

    # Per-filter value helpers
    def _dur(filter_list, i):
        return filter_list[i] if i < len(filter_list) else p.filter_duration_s

    def _margin(margin_list, i):
        return margin_list[i] if i < len(margin_list) else p.safety_margin_s

    # Current position (depth) and time
    current_depth = 0.0
    current_time = p.deploy_delay_s  # start after deck delay

    timeline.append(CastEvent("Deploy", 0.0, 0.0, "surface"))
    if p.deploy_delay_s > 0:
        timeline.append(CastEvent("Water entry", p.deploy_delay_s, 0.0, "water_entry"))

    # ------------------------------------------------------------------
    # DOWNCAST: Downcast filter stops (D_flt) (shallow → deep)
    # ------------------------------------------------------------------
    for i, depth in enumerate(t1):
        seg_time = travel_time(abs(depth - current_depth), p.v_max, p.accel)
        current_time += seg_time
        current_depth = depth

        dur = _dur(p.type1_filter_durations, i)
        margin = _margin(p.type1_safety_margins, i)
        arrival = current_time
        preset = arrival + margin
        departure = preset + p.homing_time_s + p.actuation_buffer_s + dur

        fs = FilterStop(
            filter_id=f"D-{i+1}",
            filter_type=1,
            depth_m=depth,
            arrival_time_s=arrival,
            departure_time_s=departure,
            preset_delay_s=preset,
            filter_duration_s=dur,
        )
        filter_stops.append(fs)

        timeline.append(CastEvent(f"D-{i+1} start", arrival, depth, "filter_start"))
        timeline.append(CastEvent(f"D-{i+1} end", departure, depth, "filter_end"))

        current_time = departure

    # ------------------------------------------------------------------
    # After T1 filters: return to surface before main downcast
    # ------------------------------------------------------------------
    if t1:
        seg_time = travel_time(abs(current_depth - 0.0), p.v_max, p.accel)
        current_time += seg_time
        current_depth = 0.0
        timeline.append(CastEvent("Surface return", current_time, 0.0, "surface_return"))

    # ------------------------------------------------------------------
    # DOWNCAST: Continue to bottom
    # ------------------------------------------------------------------
    seg_time = travel_time(abs(p.max_depth_m - current_depth), p.v_max, p.accel)
    current_time += seg_time
    current_depth = p.max_depth_m

    bottom_arrive = current_time
    timeline.append(CastEvent("Bottom", bottom_arrive, p.max_depth_m, "bottom"))
    bottom_leave = bottom_arrive + p.bottom_dwell_s
    if p.bottom_dwell_s > 0:
        timeline.append(CastEvent("Bottom leave", bottom_leave, p.max_depth_m, "bottom_leave"))
    current_time = bottom_leave

    # ------------------------------------------------------------------
    # UPCAST: Upcast filter stops (U_flt) (deep → shallow)
    # ------------------------------------------------------------------
    for i, depth in enumerate(t2):
        seg_time = travel_time(abs(current_depth - depth), p.v_max, p.accel)
        current_time += seg_time
        current_depth = depth

        dur = _dur(p.type2_filter_durations, i)
        margin = _margin(p.type2_safety_margins, i)
        arrival = current_time
        preset = arrival + margin
        departure = preset + p.homing_time_s + p.actuation_buffer_s + dur

        fs = FilterStop(
            filter_id=f"U-{i+1}",
            filter_type=2,
            depth_m=depth,
            arrival_time_s=arrival,
            departure_time_s=departure,
            preset_delay_s=preset,
            filter_duration_s=dur,
        )
        filter_stops.append(fs)

        timeline.append(CastEvent(f"U-{i+1} start", arrival, depth, "filter_start"))
        timeline.append(CastEvent(f"U-{i+1} end", departure, depth, "filter_end"))

        current_time = departure

    # ------------------------------------------------------------------
    # UPCAST: Recover directly to deck
    # ------------------------------------------------------------------
    seg_time = travel_time(current_depth, p.v_max, p.accel)
    current_time += seg_time
    timeline.append(CastEvent("Recovered", current_time, 0.0, "surface"))

    total_cast_time_s = current_time

    return filter_stops, timeline, total_cast_time_s


def fmt_time(seconds: float) -> str:
    """Format seconds as HH:MM:SS."""
    seconds = int(round(seconds))
    h = seconds // 3600
    m = (seconds % 3600) // 60
    s = seconds % 60
    return f"{h:02d}:{m:02d}:{s:02d}"


# ---------------------------------------------------------------------------
# Live upcast speed solver
# ---------------------------------------------------------------------------

def required_speed(distance_m: float, time_avail_s: float, accel: float):
    """
    Given a distance to travel and the time available (until the preset timer fires),
    compute the winch speed needed to arrive exactly on time.

    Uses the trapezoidal ramp model: t = v/a + d/v  →  v² - a·t·v + a·d = 0
    Minimum possible time (triangular profile): t_min = 2·√(d/a)

    Parameters
    ----------
    distance_m   : float  Distance to travel (m, positive)
    time_avail_s : float  Time available (s)
    accel        : float  Acceleration magnitude (m/s²)

    Returns
    -------
    dict with keys:
        status       : 'impossible' | 'ok'
        v_required   : float | None   Required speed (m/s) to arrive exactly on time
        t_min_s      : float          Minimum possible travel time (physics limit)
        t_at_vmax    : float | None   Travel time at current v_max (for comparison)
        early_s      : float | None   How many seconds early you'd arrive at v_max
    """
    if distance_m <= 0:
        return {"status": "ok", "v_required": 0.0, "t_min_s": 0.0,
                "t_at_vmax": 0.0, "early_s": time_avail_s}

    t_min = 2.0 * math.sqrt(distance_m / accel)

    if time_avail_s < t_min:
        return {
            "status": "impossible",
            "v_required": None,
            "t_min_s": t_min,
            "t_at_vmax": None,
            "early_s": None,
        }

    # Solve v² - a·t·v + a·d = 0  (take the smaller positive root)
    discriminant = (accel * time_avail_s) ** 2 - 4 * accel * distance_m
    v_req = (accel * time_avail_s - math.sqrt(max(discriminant, 0.0))) / 2.0

    return {
        "status": "ok",
        "v_required": v_req,
        "t_min_s": t_min,
        "t_at_vmax": None,   # filled in by the caller who knows v_max
        "early_s": None,
    }

File: docs/js/calculator.js
/* ===================================================================
   calculator.js — ChronoCaster PWA
   Exact JavaScript port of calculator.py
   =================================================================== */
'use strict';

/**
 * Travel time for a symmetric trapezoidal velocity profile.
 * @param {number} distanceM  Distance in metres (non-negative)
 * @param {number} vMax       Maximum speed m/s
 * @param {number} accel      Acceleration/deceleration m/s²
 * @returns {number} Time in seconds
 */
function travelTime(distanceM, vMax, accel) {
  if (distanceM <= 0) return 0.0;
  const rampDist = (vMax * vMax) / (2 * accel);
  const fullRampDist = 2 * rampDist;
  if (distanceM >= fullRampDist) {
    const rampTime = vMax / accel;
    const cruiseDist = distanceM - fullRampDist;
    return 2 * rampTime + cruiseDist / vMax;
  } else {
    const vPeak = Math.sqrt(accel * distanceM);
    return 2 * vPeak / accel;
  }
}

/**
 * Required winch speed to travel distanceM in timeAvailS seconds.
 * Solves v² − a·t·v + a·d = 0 (smaller positive root).
 * @returns {{status:'ok'|'impossible', vRequired:number|null, tMinS:number}}
 */
function requiredSpeed(distanceM, timeAvailS, accel) {
  if (distanceM <= 0)
    return { status: 'ok', vRequired: 0.0, tMinS: 0.0 };

  const tMin = 2.0 * Math.sqrt(distanceM / accel);
  if (timeAvailS < tMin)
    return { status: 'impossible', vRequired: null, tMinS: tMin };

  const discriminant = Math.pow(accel * timeAvailS, 2) - 4 * accel * distanceM;
  const vReq = (accel * timeAvailS - Math.sqrt(Math.max(discriminant, 0.0))) / 2.0;
  return { status: 'ok', vRequired: vReq, tMinS: tMin };
}

/**
 * Compute preset delays for all filters and build the full cast timeline.
 *
 * @param {object} p  CastParams object with fields:
 *   maxDepthM, type1Depths[], type2Depths[], vMax, accel,
 *   filterDurationS, safetyMarginS,
 *   type1FilterDurations[], type2FilterDurations[],
 *   type1SafetyMargins[], type2SafetyMargins[],
 *   deployDelayS, bottomDwellS, homingTimeS, actuationBufferS
 *
 * @returns {{ filterStops: FilterStop[], timeline: Event[], totalCastTimeS: number }}
 */
function computeDelays(p) {
  if (!p.type1Depths.length && !p.type2Depths.length)
    throw new Error('At least one filter depth must be provided.');

  const t1 = [...p.type1Depths].sort((a, b) => a - b);   // shallow → deep
  const t2 = [...p.type2Depths].sort((a, b) => b - a);   // deep → shallow

  const getDur    = (list, i) => (i < list.length) ? list[i] : p.filterDurationS;
  const getMargin = (list, i) => (i < list.length) ? list[i] : p.safetyMarginS;

  const filterStops = [];
  const timeline    = [];

  let currentDepth = 0.0;
  let currentTime  = p.deployDelayS;    // start after deck delay

  timeline.push({ label: 'Deploy', timeS: 0.0, depthM: 0.0, eventType: 'surface' });
  if (p.deployDelayS > 0)
    timeline.push({ label: 'Water entry', timeS: p.deployDelayS, depthM: 0.0, eventType: 'water_entry' });

  // ---- Downcast filter stops (T1) ----
  t1.forEach((depth, i) => {
    currentTime  += travelTime(Math.abs(depth - currentDepth), p.vMax, p.accel);
    currentDepth  = depth;

    const dur     = getDur(p.type1FilterDurations, i);
    const margin  = getMargin(p.type1SafetyMargins, i);
    const arrival   = currentTime;
    const preset    = arrival + margin;
    const departure = preset + p.homingTimeS + p.actuationBufferS + dur;

    filterStops.push({
      filterId: `D-${i + 1}`, filterType: 1, depthM: depth,
      arrivalTimeS: arrival, departureTimeS: departure,
      presetDelayS: preset, filterDurationS: dur,
    });
    timeline.push({ label: `D-${i+1} start`, timeS: arrival,    depthM: depth, eventType: 'filter_start' });
    timeline.push({ label: `D-${i+1} end`,   timeS: departure,  depthM: depth, eventType: 'filter_end'   });
    currentTime = departure;
  });

  // ---- After T1 filters: return to surface before main downcast ----
  if (t1.length > 0) {
    currentTime  += travelTime(currentDepth, p.vMax, p.accel);
    currentDepth  = 0.0;
    timeline.push({ label: 'Surface return', timeS: currentTime, depthM: 0.0, eventType: 'surface_return' });
  }

  // ---- Main downcast to bottom ----
  currentTime  += travelTime(Math.abs(p.maxDepthM - currentDepth), p.vMax, p.accel);
  currentDepth  = p.maxDepthM;

  const bottomArrive = currentTime;
  timeline.push({ label: 'Bottom', timeS: bottomArrive, depthM: p.maxDepthM, eventType: 'bottom' });

  const bottomLeave = bottomArrive + p.bottomDwellS;
  if (p.bottomDwellS > 0)
    timeline.push({ label: 'Bottom leave', timeS: bottomLeave, depthM: p.maxDepthM, eventType: 'bottom_leave' });
  currentTime = bottomLeave;

  // ---- Upcast filter stops (T2) ----
  t2.forEach((depth, i) => {
    currentTime  += travelTime(Math.abs(currentDepth - depth), p.vMax, p.accel);
    currentDepth  = depth;

    const dur     = getDur(p.type2FilterDurations, i);
    const margin  = getMargin(p.type2SafetyMargins, i);
    const arrival   = currentTime;
    const preset    = arrival + margin;
    const departure = preset + p.homingTimeS + p.actuationBufferS + dur;

    filterStops.push({
      filterId: `U-${i + 1}`, filterType: 2, depthM: depth,
      arrivalTimeS: arrival, departureTimeS: departure,
      presetDelayS: preset, filterDurationS: dur,
    });
    timeline.push({ label: `U-${i+1} start`, timeS: arrival,    depthM: depth, eventType: 'filter_start' });
    timeline.push({ label: `U-${i+1} end`,   timeS: departure,  depthM: depth, eventType: 'filter_end'   });
    currentTime = departure;
  });

  // ---- Recover to deck ----
  currentTime += travelTime(currentDepth, p.vMax, p.accel);
  timeline.push({ label: 'Recovered', timeS: currentTime, depthM: 0.0, eventType: 'surface' });

  return { filterStops, timeline, totalCastTimeS: currentTime };
}

// =====================================================================
// Formatting utilities
// =====================================================================

/** Format seconds as HH:MM:SS (handles negative with a leading −) */
function fmtTime(seconds) {
  if (isNaN(seconds) || !isFinite(seconds)) return '--:--:--';
  const neg = seconds < 0;
  seconds = Math.round(Math.abs(seconds));
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  const str = `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
  return neg ? '−' + str : str;
}

/** Parse MM:SS, M:SS, H:MM:SS, or plain seconds. Returns 0 on failure. */
function parseMMSS(text) {
  text = (text || '').trim();
  const parts = text.split(':');
  try {
    if (parts.length === 3) return parseInt(parts[0]) * 3600 + parseInt(parts[1]) * 60 + parseInt(parts[2]);
    if (parts.length === 2) return parseInt(parts[0]) * 60 + parseInt(parts[1]);
    return Math.round(parseFloat(text)) || 0;
  } catch (e) { return 0; }
}

/** Seconds → "H:MM:SS" (no zero-padding on hours) */
function sToHMMSS(s) {
  s = Math.round(Math.max(s, 0));
  return `${Math.floor(s / 3600)}:${String(Math.floor((s % 3600) / 60)).padStart(2, '0')}:${String(s % 60).padStart(2, '0')}`;
}
